src/lib.rs

```diff
@@ -14,39 +24,184 @@ fn process_instruction(
         accounts.len(),
         instruction_data
     ));
-    Ok(())
+
+    let account_info_iter = &mut accounts.iter();
+    let program_token_info = next_account_info(account_info_iter)?;
+    let (program_token_address, program_token_bump_seed) =
+        Pubkey::find_program_address(&[br"program-token"], program_id);
+
+    if program_token_address != *program_token_info.key {
+        info!("Error: program token address derivation mismatch");
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let program_token_signer_seeds: &[&[_]] = &[br"program-token", &[program_token_bump_seed]];
+
+    match instruction_data.get(0) {
+        Some(0) => {
+            info!("Create program token account...");
+            let funder_info = next_account_info(account_info_iter)?;
+            let mint_info = next_account_info(account_info_iter)?;
+            let system_program_info = next_account_info(account_info_iter)?;
+            let spl_token_program_info = next_account_info(account_info_iter)?;
+            let rent_sysvar_info = next_account_info(account_info_iter)?;
+
+            let rent = &Rent::from_account_info(rent_sysvar_info)?;
+
+            invoke_signed(
+                &system_instruction::create_account(
```

Comment:
i noticed you used `system_instruction::create_account()` here, is that any different than using `pubkey::create_program_address()` https://github.com/solana-labs/solana/blob/89b474e192ea2f4b2274ad26cbc47443d5767571/sdk/program/src/pubkey.rs#L135

Reply:
Yes, very different.  Sadly the names are close enough that I can see where confusion can arise!

`system_instruction::create_account()` - this is transaction instruction implemented by the system program that when executed allocates a new account on the chain based on the given parameters.   The address that the account is allocated at is also provided as an input to this instruction

`pubkey::create_program_address()` - a local function that returns an account address owned by the specified program, assuming the seeds are valid.  This is just an address, a 32 byte number.

Reply:
ok, and i presume you chose to use `Pubkey::find_program_address()` instead of `Pubkey::create_program_address()`, because as you mentioned above in the initial comments we have a constant address in this program for users to deposit SPL tokens

i'll try it myself, but it seems to me that i can also just call the `Pubkey::find_program_address(&[br"program-token2"], program_id);` and it'll make a second associated account users can send same tokens to

then, i can use a totally different program_id for a whole different SPL token in the same program and be able to intake 2 SPL tokens, right?

Reply:
and after creating them, this program can do invoke_signed on them and send them wherever they want

Reply:
Yep, that's right.

>  i can use a totally different program_id for a whole different SPL token in this same program and be able to intake 2 SPL tokens: Pubkey::find_program_address(&[br"program-token"], program_id2); right?

Alternatively you use the same program, but adding more seeds to `find_program_address`:
```
Pubkey::find_program_address(&[br"program-token", some_other_seed], program_id);
```
Here's an example to check out: https://github.com/solana-labs/solana-program-library/blob/487ad2d2d7cb0b8a58f02a156a659e6c8a74f354/associated-token-account/program/src/lib.rs#L24-L32

Reply:
thank you, this is very helpful

Reply:
is `program_id` in the `Pubkey::find_program_address()` function supposed to be the SPL token address that i've created (not the `spl_token::id()`? for example, I've created an SPL token, and i want to create an associated account for that token for my program (same as the one above)